Sources/main.cpp

```diff
@@ -101,6 +104,120 @@ void GetArchiveFileList(const char *path,
     }
 }
 
+class MalformedMappingFileError : public std::exception
+{
+public:
+    explicit MalformedMappingFileError(off_t lineNumber)
+        : lineNumber(lineNumber)
+    {
+        this->SetFileName(nullptr);
+    }
+
+    const char *what() const noexcept override
+    {
+        return this->message.c_str();
+    }
+
+    void SetFileName(const char *fileName)
+    {
+        if (!fileName || strcmp(fileName, "") == 0) {
+            fileName = "(unknown)";
+        }
+        std::ostringstream ss;
+        ss << "Malformed mapping file: " << fileName << ":" << this->lineNumber;
+        this->message = ss.str();
+    }
+
+private:
+    std::string message;
+    off_t lineNumber;
+};
+
+void GetArchiveFileListFromMappingFile(
+    std::istream &mappingFile,
+    std::unordered_map<std::string, std::string> &fileNames)
+{
+    static const char kWhitespace[] = " \t";
+    // TODO(strager): Make this parser more accepting. This parser is way too
+    // strict.
+    bool didReadHeader = false;
+    off_t lineNumber = 1;
+    while (mappingFile.good()) {
+        std::string line;
+        std::getline(mappingFile, line, '\n');
+        if (mappingFile.fail() && !line.empty()) {
```

Comment:
eofbit would be set in the case of a too-long line, so that's counterproductive to this line. =] Also, eofbit would be set for an input without a trailing newline.

I'll add more tests and play with things though.